include/mata/synchronized_iterator.hh

```diff
@@ -0,0 +1,260 @@
+        // needed in computation of post
+        // in subset construction, product, and non-determinization.
+        // Key is the type stored in OrdVectors, it must be comparable with <,>,==,!=,<=,>=,
+        // and it must be a total (linear) ordering.
+        // The intended usage in, for instance, determinisation is for Key to be Move.
+        // Move is ordered by the symbol.
+        //
+        // SyncrhonisedIterator is the parent virtual class.
+        // It stores a vector of end-iterators for the OrdVectors and a vector of current positions.
+        // They are filled in using the function push_back(v), that adds v.begin() to positions and v.end() to ends.
+        // Method advance advances all positions forward so that they are synchronized on the next smallest equiv class
+        // (next smallest symbol in the case of Move).
+        //
+        // There are two versions of the class.
+        // i) In product, ALL positions must point to currently the smallest equiv. class (Moves with the same symbol).
+        // Method get_current then returns the vector of all position iterators, synchronized.
+        // In determinization, it is enough that there EXISTS a position that points to the smallest class.
+        // Method get_current then returns the vector of only those positions that point to the smallest equiv. class.
+        //
+        // Usage: 0) construct, 1) fill in using push_back, iterate using advance and get_current, 2) reset, goto 1)
+
+        template<typename Key> class SynchronizedIterator {
+        public:
+            using Iterator = typename OrdVector<Key>::const_iterator;
+
+            std::vector<Iterator> positions;
+            std::vector<Iterator> ends;
+
+            explicit SynchronizedIterator(const int size = 0)
+            {
+                positions.reserve(size);
+                ends.reserve(size);
+            };
+
+            // This is supposed to be called only before an iteration,
+            // after constructor of reset.
+            // Calling after advance breaks the iterator.
+            virtual void push_back (const OrdVector<Key> &vector) {
+                // hope that the parameter is not deep copied ... is it?
+                this->positions.push_back(vector.begin());
+                this->ends.push_back(vector.end());
+            };
+
+            // Empties positions and ends.
+            // Though they should keep the allocated space.
+            void reset(const int size = 0) {
+                positions.clear();
+                ends.clear();
+                if (size > 0) {
+                    positions.reserve(size);
+                    ends.reserve(size);
+                }
+            };
+
+            virtual bool advance() = 0;
+            virtual const std::vector<Iterator> get_current() = 0;
+        };
+
+
+
+        template<typename Key>
+        class SynchronizedUniverzalIterator: public SynchronizedIterator<Key> {
```

Comment:
is Univerzal an english word?

Reply:
Do you mean that z should be s or something else? 
Indeed seem that z should be s.